nixos/modules/services/web-apps/zitadel.nix

```
@@ -0,0 +1,223 @@
+          };
+        '';
+        description = mdDoc ''
+          Contents of the runtime configuration file. See
+          https://zitadel.com/docs/self-hosting/manage/configure for more
+          details.
+        '';
+      };
+
+      extraSettingsPaths = mkOption {
+        type = types.listOf types.path;
+        default = [ ];
+        description = mdDoc ''
+          A list of paths to extra settings files. These will override the
+          values set in [settings](#opt-services.zitadel.settings). Useful if
+          you want to keep sensitive secrets out of the Nix store.
+        '';
+      };
+
+      steps = mkOption {
+        type = settingsFormat.type;
+        default = { };
+        example = lib.literalExpression ''
+          {
+            FirstInstance = {
+              InstanceName = "Example";
+              Org.Human = {
+                UserName = "foobar";
+                FirstName = "Foo";
+                LastName = "Bar";
+              };
+            };
+          }
+        '';
+        description = mdDoc ''
+          Contents of the database initialization config file. See
+          https://zitadel.com/docs/self-hosting/manage/configure for more
+          details.
+        '';
+      };
+
+      extraStepsPaths = mkOption {
+        type = types.listOf types.path;
+        default = [ ];
+        description = mdDoc ''
+          A list of paths to extra steps files. These will override the values
+          set in [steps](#opt-services.zitadel.steps). Useful if you want to
+          keep sensitive secrets out of the Nix store.
+        '';
+      };
+    };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [{
+      assertion = cfg.tlsMode != "enabled"
+        || ((cfg.settings.TLS.Key != null || cfg.settings.TLS.KeyPath != null)
+          && (cfg.settings.TLS.Cert != null || cfg.settings.TLS.CertPath
+            != null));
+      message = ''
+        A TLS certificate and key must be configured in
+        services.zitadel.settings.TLS if services.zitadel.tlsMode is enabled.
+      '';
+    }];
+
+    networking.firewall.allowedTCPPorts =
+      lib.mkIf cfg.openFirewall [ cfg.settings.Port ];
+
+    systemd.services.zitadel = let
+      configFile = settingsFormat.generate "config.yaml" cfg.settings;
+      stepsFile = settingsFormat.generate "steps.yaml" cfg.steps;
+
+      mkArgs = name: args:
+        builtins.concatStringsSep " " (map (f: "--${name} ${f}") args);
+      configArgs = mkArgs "config" (cfg.extraSettingsPaths ++ [ configFile ]);
+      stepsArgs = mkArgs "steps" (cfg.extraStepsPaths ++ [ stepsFile ]);
+    in {
+      description = "ZITADEL identity access management";
+      path = [ cfg.package ];
+      wantedBy = [ "multi-user.target" ];
+
+      script = ''
+        zitadel start-from-init \
+          ${configArgs} \
+          ${stepsArgs} \
+          --masterkeyFile ${cfg.masterKeyFile} \
+          --tlsMode ${cfg.tlsMode}
+      '';
```

I think you can simplify some of this using `lib.toGNUCommandLine`, like this:

```suggestion
    in {
      description = "ZITADEL identity access management";
      path = [ cfg.package ];
      wantedBy = [ "multi-user.target" ];

      script = ''
        zitadel start-from-init ${lib.toGNUCommandLine { } {
          config = cfg.extraSettingsPaths ++ [ configFile ];
          steps = cfg.extraSettingsPaths ++ [ stepsFile ];
          masterkeyFile = cfg.masterKeyFile;
          inherit (cfg) tlsMode;
      }'';
```